# Quero analisar os dados de passageiros do Titanic, disponíveis como arquivo CSV.


# #O pandas fornece a função read_csv() para ler dados armazenados como um arquivo csv em um pandas DataFrame.
## O pandas suporta muitos formatos de arquivo diferentes ou fontes de dados prontas para uso (csv, excel, sql, json, parquet, …), cada um deles com o prefixo read_*.

In [1]:
import pandas as pd
from matplotlib import pyplot as plt
titanic = pd.read_csv("file:///home/dcdevs/Documents/IFTM/ADS/ads-section-2/arquiteturaOrgComputadores/Python/titanic.csv")

In [2]:
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

# Eu quero ver as primeiras 8 linhas de um DataFrame pandas

In [3]:
titanic.head(8)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S

# Para ver as primeiras N linhas de a DataFrame, use o head()método com o número necessário de linhas (neste caso 8) como argumento.

In [4]:
# Em vez disso, está interessado nas últimas N linhas? 
# pandas também fornece um tail()método. Por exemplo, titanic.tail(10)retornará as últimas 10 linhas do DataFrame.
titanic.tail(10)

PassengerId  Survived  Pclass                                      Name  \
881          882         0       3                        Markun, Mr. Johann   
882          883         0       3              Dahlberg, Miss. Gerda Ulrika   
883          884         0       2             Banfield, Mr. Frederick James   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
881    male  33.0      0      0            349257   7.8958   NaN        S  
882  female  22.0      0      0              7552  10.5167   NaN        S  
883    male  28.0      0      0  C.A./SOTON 34068  10.5000   NaN        S  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5            382652  29.1250   NaN        Q  
886    male  27.0      0      0            211536  13.0000   NaN        S  
887  female  19.0      0      0            112053  30.0000   B42        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
889    male  26.0      0      0            111369  30.0000  C148        C  
890    male  32.0      0      0            370376   7.7500   NaN        Q

In [5]:
#Uma verificação de como os
#pandas interpretaram cada um dos tipos de dados
#da coluna pode ser feita solicitando o dtypesatributo pandas:
titanic.dtypes
#Para cada uma das colunas, o tipo de dados usado é listado. 
#Os tipos de dados DataFramesão inteiros ( int64), floats ( float64) e strings ( object).

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

Ao pedir o dtypes, nenhum colchete é usado! dtypesé um atributo de a DataFramee Series. Atributos de DataFrameou Seriesnão precisam de colchetes. Atributos representam uma característica de um DataFrame/ Series, enquanto um método (que requer colchetes) faz algo com o DataFrame/ Seriesconforme apresentado no primeiro tutorial .

In [8]:
#Meu colega solicitou os dados do Titanic como uma planilha.

#titanic.to_excel("file:///home/dcdevs/Documents/IFTM/ADS/ads-section-2/arquiteturaOrgComputadores/Python/titanic.xlsx", sheet_name="passageiros", index=False)

# Enquanto read_*as funções são usadas para ler dados para pandas, os to_*métodos são usados ​​para armazenar dados. O to_excel()método armazena os dados como um arquivo excel. No exemplo aqui, o sheet_namenome é passageiros em vez do Sheet1 padrão . Ao definir index=Falseos rótulos de índice de linha não são salvos na planilha
A função de leitura equivalente read_excel()recarregará os dados para um DataFrame

In [10]:
#titanic = pd.read_excel("file:///home/dcdevs/Documents/IFTM/ADS/ads-section-2/arquiteturaOrgComputadores/Python/titanic.xlsx", sheet_name="passageiros")

In [11]:
#Estou interessado em um resumo técnico de umDataFrame
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


# O método info()fornece informações técnicas sobre um DataFrame


# Como seleciono um subconjunto de um DataFrame?
## Estou interessado na idade dos passageiros do Titanic.

In [12]:
ages = titanic["Age"]
ages.head()

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: Age, dtype: float64

# Cada coluna em a DataFrame é um Series. Como uma única coluna é selecionada, o objeto retornado é um pandas Series. Podemos verificar isso verificando o tipo da saída:

In [13]:
type(titanic["Age"])

pandas.core.series.Series

In [14]:
#E dê uma olhada na shapesaída:
titanic["Age"].shape
#DataFrame.shapeé um atributo de um pandas Seriese DataFrame 
#contendo o número de linhas e colunas: (nrows, ncolumns) . 
#Uma série de pandas é unidimensional e apenas o número de linhas é retornado.

(891,)

In [22]:
age_sex = titanic[["Age", "Sex", "Name"]]
age_sex.head()

Age     Sex                                               Name
0  22.0    male                            Braund, Mr. Owen Harris
1  38.0  female  Cumings, Mrs. John Bradley (Florence Briggs Th...
2  26.0  female                             Heikkinen, Miss. Laina
3  35.0  female       Futrelle, Mrs. Jacques Heath (Lily May Peel)
4  35.0    male                           Allen, Mr. William Henry

Como faço para filtrar linhas específicas de um DataFrame?

In [16]:
above_35 = titanic[titanic["Age"] > 35]
above_35.head()

PassengerId  Survived  Pclass  \
1             2         1       1   
6             7         0       1   
11           12         1       1   
13           14         0       3   
15           16         1       2   

                                                 Name     Sex   Age  SibSp  \
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   
13                        Andersson, Mr. Anders Johan    male  39.0      1   
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   

    Parch    Ticket     Fare Cabin Embarked  
1       0  PC 17599  71.2833   C85        C  
6       0     17463  51.8625   E46        S  
11      0    113783  26.5500  C103        S  
13      5    347082  31.2750   NaN        S  
15      0    248706  16.0000   NaN        S

Para selecionar linhas com base em uma expressão condicional,
use uma condição dentro dos colchetes de seleção [].
A condição dentro dos colchetes de seleção verifica para quais 
linhas a coluna tem um valor maior que 35:titanic["Age"] > 35Age

In [17]:
titanic["Age"] > 35

0      False
1       True
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Age, Length: 891, dtype: bool

## A saída da expressão condicional ( >, mas também ==, !=, <, <=,… funcionaria) é na verdade um pandas Seriesde valores booleanos (ou Trueou False) com o mesmo número de linhas que o original DataFrame. Esses Seriesvalores booleanos podem ser usados ​​para filtrar o DataFramecolocando-o entre os colchetes de seleção []. Somente as linhas para as quais o valor é True serão selecionadas.

## Sabemos de antes que o Titanic original DataFrame consiste em 891 linhas. Vamos dar uma olhada no número de linhas que satisfazem a condição verificando o atributo shape do resultado DataFrame above_35:

In [18]:
above_35.shape

(217, 12)

Estou interessado nos passageiros do Titanic das classes de cabine 2 e 3.

In [19]:
class_23 = titanic[titanic["Pclass"].isin([2, 3])]

class_23.head()

PassengerId  Survived  Pclass                            Name     Sex  \
0            1         0       3         Braund, Mr. Owen Harris    male   
2            3         1       3          Heikkinen, Miss. Laina  female   
4            5         0       3        Allen, Mr. William Henry    male   
5            6         0       3                Moran, Mr. James    male   
7            8         0       3  Palsson, Master. Gosta Leonard    male   

    Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0  22.0      1      0         A/5 21171   7.2500   NaN        S  
2  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
4  35.0      0      0            373450   8.0500   NaN        S  
5   NaN      0      0            330877   8.4583   NaN        Q  
7   2.0      3      1            349909  21.0750   NaN        S

Semelhante à expressão condicional, a isin()função condicional retorna um Truepara cada linha em que os valores estão na lista fornecida. Para filtrar as linhas com base nessa função, use a função condicional dentro dos colchetes de seleção []. Nesse caso, a condição dentro dos colchetes de seleção verifica para quais linhas a coluna é 2 ou 3.titanic["Pclass"].isin([2, 3]) Pclass

O acima é equivalente a filtrar por linhas para as quais a classe é 2 ou 3 e combinar as duas instruções com um |operador (or):

In [20]:
class_23 = titanic[(titanic["Pclass"] == 2) | (titanic["Pclass"] == 3)]

class_23.head()

PassengerId  Survived  Pclass                            Name     Sex  \
0            1         0       3         Braund, Mr. Owen Harris    male   
2            3         1       3          Heikkinen, Miss. Laina  female   
4            5         0       3        Allen, Mr. William Henry    male   
5            6         0       3                Moran, Mr. James    male   
7            8         0       3  Palsson, Master. Gosta Leonard    male   

    Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0  22.0      1      0         A/5 21171   7.2500   NaN        S  
2  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
4  35.0      0      0            373450   8.0500   NaN        S  
5   NaN      0      0            330877   8.4583   NaN        Q  
7   2.0      3      1            349909  21.0750   NaN        S